src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1282,7 +1265,25 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
     }
 
     void visit_Return(const AST::Return_t &x) {
-        tmp = ASR::make_Return_t(al, x.base.base.loc);
+        std::string return_var_name = "_lpython_return_variable";
+        LFORTRAN_ASSERT(current_scope->scope.find(return_var_name) != current_scope->scope.end())
+        this->visit_expr(*x.m_value);
+        ASR::expr_t *value = ASRUtils::EXPR(tmp);
+        ASR::symbol_t *return_var = current_scope->scope[return_var_name];
+        ASR::asr_t *return_var_ref = ASR::make_Var_t(al, x.base.base.loc, return_var);
+        ASR::expr_t *target = ASRUtils::EXPR(return_var_ref);
+        ASR::ttype_t *target_type = ASRUtils::expr_type(target);
+        ASR::ttype_t *value_type = ASRUtils::expr_type(value);
+        if (!ASRUtils::check_equal_type(target_type, value_type)) {
+            std::string ltype = ASRUtils::type_to_str(target_type);
+            std::string rtype = ASRUtils::type_to_str(value_type);
+            throw SemanticError("Type Mismatch in return, found (" +
+                    ltype + " and " + rtype + ")", x.base.base.loc);
+        }
+
+        ASR::stmt_t *overloaded=nullptr;
+        tmp = ASR::make_Assignment_t(al, x.base.base.loc, target, value,
+                                overloaded);
```

Comment:
Everything looks correct to me in this PR, except here after the assignment we also have to emit Return in ASR, see my other comment in the C++ generated code.

The question becomes how to do that, since our visitor is assuming that every AST node transforms to 0 or 1 ASR nodes, but here we need to emit 2 ASR nodes. I think that the cleanest way is to save the `body` Vector as an internal variable, and then insert the instruction from here.

Reply:
Fixed now.